include/deal.II/fe/fe_coupling_values.h

```diff
@@ -0,0 +1,641 @@
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_fe_coupling_values_h
+#define dealii_fe_coupling_values_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/algorithms/general_data_storage.h>
+
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/thread_local_storage.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/fe/fe_values_views.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declaration
+#ifndef DOXYGEN
+template <int dim, int spacedim>
+class FEValuesBase;
+#endif
+
+namespace FEValuesViews
+{
+  /**
+   * A class that stores the renumbering of degrees of freedom and quadrature
+   * points for the RenumberedView class. This class is common to all possible
+   * renumbered views, and therefore it can be shared between all views that use
+   * the same set of renumberings.
+   *
+   * The renumbering is stored in two vectors, one for the degrees of freedom
+   * and one for the quadrature points. The renumbering is applied by
+   * FEValuesViews::ReorderedView.
+   */
+  struct RenumberingData
+  {
+    /**
+     * Construct a new renumbering data object.
+     *
+     * The @p dof_renumbering vector is used to renumber the degrees of freedom,
+     * while the @p quadrature_renumbering vector is used to renumber the
+     * quadrature points. An empty renumbering vector simply means that no
+     * renumbering is performed.
+     *
+     * @note The renumbering vectors are *not* required to match the dimensions
+     * of the underlying view, i.e., you could use this class to only run over
+     * half of the degrees of freedom, and integrating only over half of the
+     * current cell by selecting a subset of quadrature points, if you wish to
+     * do so, or to run over more degrees of freedom w.r.t. to the underlying
+     * view. The important part is that every entry of each renumbering vector
+     * is a legal value for the underlying view object. We allow the dof
+     * renumbering vector to contain `numbers::invalid_unsigned_int` values,
+     * which are ignored, and produce zero values, gradients, hessians, etc. for
+     * the corresponding shape function.
+     *
+     * An example of the renumbering ordering is the following. Assume that you
+     * have an FE_Q(1) finite element spac in dimension two (i.e., four degrees
+     * of freedom), and that you are manually taking care of an additional
+     * degree of freedom (say, a custom shape function in the middle of the
+     * cell), which, for your own convenience, should be numbered locally in the
+     * middle of the others. You would like your loops on degrees of freedom to
+     * run over all degrees of freedom, and to have the same structure as if the
+     * additional degree of freedom was not there. Then the renumbering data
+     * object should look like this:
+     * @code
+     * ...
+     * RenumberingData data({{0, 1, numbers::invalid_unsigned_int, 2, 3}});
```

Comment:
But could it also be {0,1,2,3}?

Reply:
Yes. In this case it would simply be the identity renumbering.